src/bitcoind_client.rs

```diff
@@ -0,0 +1,188 @@
+//! This module implements a minimal and non standard conforming HTTP client that works with
+//! the bitcoind RPC server. This client can be used if minimal dependencies are a goal.
+
+use ::client::HttpRoundTripper;
+
+use http::{Request, Response};
+
+use std;
+use std::io::{BufRead, BufReader, Cursor, Write};
+use std::net::TcpStream;
+use std::time::{Instant, Duration};
+
+/// Simple bitcoind JSON RPC client that implements the necessary subset of HTTP
+#[derive(Copy, Clone, Debug)]
+pub struct SimpleBitcoindClient {
+    default_port: u16,
+    timeout: Duration,
+}
+
+/// Builder for non-standard `SimpleBitcoinClient`s
+#[derive(Clone, Debug)]
+pub struct SimpleBitcoindClientBuilder {
+    client: SimpleBitcoindClient,
+}
+
+impl Default for SimpleBitcoindClient {
+    fn default() -> Self {
+        SimpleBitcoindClient {
+            default_port: 8332,
+            timeout: Duration::from_secs(15),
+        }
+    }
+}
+
+
+impl SimpleBitcoindClientBuilder {
+    /// Construct new `SimpleBitcoinClientBuilder`
+    pub fn new() -> SimpleBitcoindClientBuilder {
+        SimpleBitcoindClientBuilder {
+            client: SimpleBitcoindClient::new(),
+        }
+    }
+
+    /// Sets the port that the client will connect to in case none was specified in the URL of the
+    /// request.
+    pub fn default_port(&mut self, port: u16) -> &mut Self {
+        self.client.default_port = port;
+        self
+    }
+
+    /// Sets the timeout after which requests will abort if they aren't finished
+    pub fn timeout(&mut self, timeout: Duration) -> &mut Self {
+        self.client.timeout = timeout;
+        self
+    }
+
+    /// Builds the final `SimpleBitcoindClient`
+    pub fn build(self) -> SimpleBitcoindClient {
+        self.client
+    }
+}
+
+impl SimpleBitcoindClient {
+    /// Construct a new `SimpleBitcoindClient` with default parameters
+    pub fn new() -> Self {
+        SimpleBitcoindClient::default()
+    }
+
+    /// Returns a builder for `SimpleBitcoindClient`
+    pub fn builder() -> SimpleBitcoindClientBuilder {
+        SimpleBitcoindClientBuilder::new()
+    }
+}
+
+/// Try to read a line from a buffered reader. If no line can be read till the deadline is reached
+/// return a timeout error.
+fn get_line<R: BufRead>(reader: &mut R, deadline: Instant) -> Result<String, Error> {
+    let mut line = String::new();
+    while deadline > Instant::now() {
+        match reader.read_line(&mut line) {
```

Comment:
This only works if `reader` is non-blocking? And I can't really tell where in `SimpleBitcoindClient::post` `TcpSocket` is set to non-blocking.

Reply:
Good catch. Yeah, this method looks really weird -- the call to `yield_now` suggests this is supposed to be run in its own thread, but even then, a blocking socket means that the `while` potentially won't ever be reached. There is also a bunch of timeout logic in the code which is used to set `deadline` but which never affects the underlying socket.

Let me see if I can add some unit tests for this.

Reply:
Lol, upon actually testing this there are a ton of ergonomic issues with the API. I will add some more commits.

Reply:
So, for what it's worth, the timeout logic seems to work when testing. I'm just not sure how :)

Reply:
@sgeisler can you offer some insight as to why the timeout code works? I'm really at a loss here, after reading various docs and some of the libstd source I don't see why the first call to `read_line` won't block indefinitely.

Reply:
Is some call to https://doc.rust-lang.org/std/net/struct.TcpStream.html#method.set_nonblocking hiding somewhere?